File: app/src/main/java/com/edotassi/amazmod/adapters/DonateProductsAdapter.java
package com.edotassi.amazmod.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.SkuDetails;
import com.edotassi.amazmod.Interface.IDonateProductClickListener;
import com.edotassi.amazmod.R;
import com.edotassi.amazmod.ui.DonationActivity;

import java.util.List;

public class DonateProductsAdapter extends RecyclerView.Adapter<DonateProductsAdapter.MyViewHolder> {

    DonationActivity donationActivity;
    List<SkuDetails> skuDetailsList;
    BillingClient billingClient;

    public DonateProductsAdapter(DonationActivity donationActivity, List<SkuDetails> skuDetailsList, BillingClient billingClient) {
        this.donationActivity = donationActivity;
        this.skuDetailsList = skuDetailsList;
        this.billingClient = billingClient;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(donationActivity.getBaseContext())
                .inflate(R.layout.row_donate_item, parent,false);
        return new MyViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        holder.textProductName.setText(skuDetailsList.get(position).getTitle());
        holder.textProductDescription.setText(skuDetailsList.get(position).getDescription());
        holder.textProductPrice.setText(skuDetailsList.get(position).getPrice());

        holder.setiDonateProductClickListener(new IDonateProductClickListener() {
            @Override
            public void onProductClickListener(View view, int position) {
                BillingFlowParams billingFlowParams = BillingFlowParams.newBuilder()
                        .setSkuDetails(skuDetailsList.get(position))
                        .build();
                billingClient.launchBillingFlow(donationActivity, billingFlowParams);
            }
        });

    }

    @Override
    public int getItemCount() {
        return skuDetailsList.size();
    }

    public class MyViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView textProductName;
        TextView textProductDescription;
        TextView textProductPrice;

        IDonateProductClickListener iDonateProductClickListener;

        public void setiDonateProductClickListener(IDonateProductClickListener iDonateProductClickListener) {
            this.iDonateProductClickListener = iDonateProductClickListener;
        }

        public MyViewHolder(@NonNull View itemView) {
            super(itemView);
            textProductName = (TextView)itemView.findViewById(R.id.row_donate_product_name);
            textProductDescription = (TextView)itemView.findViewById(R.id.row_donate_product_description);
            textProductPrice = (TextView)itemView.findViewById(R.id.row_donate_product_price);
            itemView.setOnClickListener(this);
        }


        @Override
        public void onClick(View v) {
            iDonateProductClickListener.onProductClickListener(v,getAdapterPosition());
        }
    }
}

File: app/src/main/java/com/edotassi/amazmod/adapters/DonateProductsAdapter.kt
package com.edotassi.amazmod.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.SkuDetails
import com.edotassi.amazmod.Interface.IDonateProductClickListener
import com.edotassi.amazmod.R
import com.edotassi.amazmod.ui.DonationActivity

class DonateProductsAdapter(var donationActivity: DonationActivity, var skuDetailsList: List<SkuDetails>, var billingClient: BillingClient) : RecyclerView.Adapter<DonateProductsAdapter.MyViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val view = LayoutInflater.from(donationActivity).inflate(R.layout.row_donate_item, parent, false)
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.textProductName.text = skuDetailsList[position].title
        holder.textProductDescription.text = skuDetailsList[position].description
        holder.textProductPrice.text = skuDetailsList[position].price
        holder.setiDonateProductClickListener(IDonateProductClickListener { _, p ->
            val billingFlowParams = BillingFlowParams.newBuilder()
                    .setSkuDetails(skuDetailsList[p])
                    .build()
            billingClient.launchBillingFlow(donationActivity, billingFlowParams)
        })
    }

    override fun getItemCount(): Int {
        return skuDetailsList.size
    }

    inner class MyViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        var textProductName: TextView = view.findViewById(R.id.row_donate_product_name)
        var textProductDescription: TextView = view.findViewById(R.id.row_donate_product_description)
        var textProductPrice: TextView = view.findViewById(R.id.row_donate_product_price)
        var iDonateProductClickListener: IDonateProductClickListener? = null
        fun setiDonateProductClickListener(iDonateProductClickListener: IDonateProductClickListener?) {
            this.iDonateProductClickListener = iDonateProductClickListener
        }

        override fun onClick(v: View) {
            iDonateProductClickListener!!.onProductClickListener(v, adapterPosition)
        }

        init {
            view.setOnClickListener(this)
        }
    }
}